concent_api/core/models.py

```diff
@@ -73,3 +77,236 @@ class MessageAuth(Model):
+            SubtaskState.FORCING_ACCEPTANCE.name,  # pylint: disable=no-member
+            SubtaskState.ADDITIONAL_VERIFICATION.name,  # pylint: disable=no-member
+        ],
+        SubtaskState.FAILED.name: [  # pylint: disable=no-member
+            SubtaskState.FORCING_REPORT.name,  # pylint: disable=no-member
+            SubtaskState.FORCING_RESULT_TRANSFER.name,  # pylint: disable=no-member
+            SubtaskState.ADDITIONAL_VERIFICATION.name,  # pylint: disable=no-member
+        ],
+    }
+
+    class Meta:
+        unique_together = (
+            ('requestor', 'task_id'),
+            ('requestor', 'subtask_id'),
+        )
+
+    task_id                     = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+    subtask_id                  = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH, unique = True)
+
+    # Relation to concent client in requestor context.
+    requestor                   = models.ForeignKey(Client, related_name = 'subtasks_as_requestor')
+
+    # Relation to concent client in provider context.
+    provider                    = models.ForeignKey(Client, related_name = 'subtasks_as_provider')
+
+    state                       = models.CharField(max_length = 32, choices = SubtaskState.choices())
+    _current_state              = None
+
+    # If in an active state, it's the time at which Concent automatically transitions to a passive state if it does
+    # not get the information it expects from the client (a timeout). Must be NULL in a passive state.
+    next_deadline               = models.DateTimeField(blank = True, null = True)
+
+    # Related messages
+    task_to_compute             = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_task_to_compute')
+    report_computed_task        = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_report_computed_task')
+    force_get_task_result       = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_force_get_task_result')
+    ack_report_computed_task    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_ack_report_computed_task')
+    reject_report_computed_task = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_reject_report_computed_task')
+    subtask_results_accepted    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_accepted')
+    subtask_results_rejected    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_rejected')
+
+    def __init__(self, *args, **kwargs):
+        super(Subtask, self).__init__(*args, **kwargs)
+        if self.pk is not None:
+            self._current_state = self.state
+
+    def clean(self):
+        super(Subtask, self).clean()
```

Comment:
@cameel Please look what validation is missing, definitely some but maybe you have it thought over.